packages/starlight/utils/git.ts

```diff
@@ -39,3 +40,76 @@ export function listGitTrackedFiles(directory: string): string[] {
 	const output = result.stdout.trim();
 	return output.split('\n').map((file) => join(directory, file));
 }
+
+function getRepoRoot(directory: string): string {
+	const result = spawnSync('git', ['rev-parse', '--show-toplevel'], {
+		cwd: directory,
+		encoding: 'utf-8',
+	});
+
+	if (result.error) {
+		return directory;
+	}
+
+	try {
+		return realpathSync(result.stdout.trim());
+	} catch {
+		return directory;
+	}
+}
+
+export function getAllNewestCommitDate(directory: string): [string, number][] {
+	const repoRoot = getRepoRoot(directory);
+
+	const gitLog = spawnSync(
+		'git',
+		[
+			'log',
+			// Format each history entry as t:<seconds since epoch>
+			'--format=t:%ct',
+			// In each entry include the name and status for each modified file
+			'--name-status',
+			'--',
+			directory,
+		],
+		{
+			cwd: repoRoot,
+			encoding: 'utf-8',
+		}
+	);
+
+	if (gitLog.error) {
+		return [];
+	}
+
+	let runningDate = Date.now();
+	const latestDates = new Map<string, number>();
+
+	for (const logLine of gitLog.stdout.split('\n')) {
+		if (logLine.startsWith('t:')) {
+			// t:<seconds since epoch>
+			runningDate = Number.parseInt(logLine.slice(2)) * 1000;
+		}
+
+		// - Added files take the format `A\t<file>`
+		// - Modified files take the format `M\t<file>`
+		// - Deleted files take the format `D\t<file>`
+		// - Renamed files take the format `R<count>\t<old>\t<new>`
+		// - Copied files take the format `C<count>\t<old>\t<new>`
+		// The name of the file as of the commit being processed is always
+		// the last part of the log line.
+		const tabSplit = logLine.lastIndexOf('\t');
+		if (tabSplit === -1) continue;
+		const fileName = logLine.slice(tabSplit + 1);
+
+		const currentLatest = latestDates.get(fileName) || 0;
+		latestDates.set(fileName, Math.max(currentLatest, runningDate));
+	}
+
+	return Array.from(latestDates.entries()).map(([file, date]) => {
+		const fileFullPath = resolve(repoRoot, file);
+		const fileInDirectory = relative(directory, fileFullPath);
+
+		return [fileInDirectory, date];
```

Comment:
To keep consistency with `getNewestCommitDate()` and the function name, would it make more sense to actually create the dates here instead of in the virtual module?

Reply:
It will have to be turned into a number to be serialized in the source and turned back into a Date inside the virtual module, so I don't think there is much advantage in that.

The `getNewestCommitDate` is used directly when using the dev server, so it returns the final value.